app/utility/base_planning_svc.py

```diff
@@ -80,8 +80,13 @@ async def remove_completed_links(operation, agent, links):
         """
         completed_links = [(l.command_hash if l.command_hash else l.command) for l in operation.chain
                            if l.paw == agent.paw and (l.finish or l.can_ignore())]
+
+        lat = [BasePlanningService._fil(k, 'remote.host.fqdn') for k in operation.chain if k.status != 1]
+        lat_links = [(x.command_hash if x.command_hash else x.command) for x in lat if x]
+
         return [l for l in links if l.ability.repeatable or
-                (l.command_hash if l.command_hash else l.command) not in completed_links]
+                ((l.command_hash if l.command_hash else l.command) not in completed_links
+                 and (l.command_hash if l.command_hash else l.command not in lat_links))]
 
```

Comment:
So here we are preventing 'remote.host.fqdn' links that have been retrieved by other agents from being returned. And these prevented links are ones that are outstanding, i.e. on the operational chain but not completed yet?

Reply:
No, it's filtering any matching links that haven't failed - so anything in progress or completed is removed for the moment.

Reply:
(reposting here form slack chat for reference)

The logic here and in get_links() for filtering out links with 'remote.host.fqdn' look similar. Can they both use same logic and have it be in a new function? (would also make me feel better about remove_completed_links() doing something other than just removing links, without any overt sign (i.e. submethod call)
 of doing so.)

Reply:
(following up from slack conversation for reference)

They filter for two different things. get_links() can only see the commands sent to agents in a single server 'tick', while remove_completed_links() can only see links per agent in context of all links previously run. With filters in both of them, we can be confident of both historical filtering, and avoiding the condition that arises when multiple agents try to run the same command at the same time having gotten through that filtering.